api/envoy/extensions/http/stateful_session/header/v3/header.proto

```diff
@@ -39,4 +39,8 @@ message HeaderBasedSessionState {
   // The name that will be used to obtain header value from downstream HTTP request or generate
   // new header for downstream.
   string name = 1 [(validate.rules).string = {min_len: 1}];
+
+  // If set to True, the HTTP request must be routed to the requested destination.
+  // If the requested destination is not available, Envoy returns 503. Defaults to False.
```

Comment:
nit: If the request destination is not available, Envoy will treat that as no healthy upstream and response with 503 status will be returned.

Reply:
A high-level question: is this valid only for header, or will this also be used for other types of stateful-session (e.g., cookie). If the latter, then I propose creating a "common" proto, that they all share, and include that in each of them.
Also, consider clarifying what does "not available" means (i.e., non-existent because it was removed, or unhealthy because of active/passive health-check).

Reply:
cc @adisuissa here is previous response from @cpakulski to this question in another issue. https://github.com/envoyproxy/envoy/issues/30475#issuecomment-1779718364

Reply:
@wbpcode That comment seems to be an argument why this should be configured in the filter instead of in the cluster, and I agree that doing it at the filter level makes sense.  But I don't think that answers the question @adisuissa raised about whether this is specific to header-based SSA or could also apply to cookie-based SSA.  It seems like it should apply the same to either mechanism -- regardless of whether the affinity information is stored in a header or a cookie, we may want to say that if the specified host is not available, we fail the request instead of picking a different host.

Perhaps this field should be moved to the stateful session filter config itself, so that it applies regardless of what session state extension is used?

Reply:
I believe that `strict` setting does not make sense for cookie-based SSA. Let's imagine that there is `strict` setting for cookie-based SSA. Envoy receives a request, determines that the desired host is not not available and returns 503. A client must send another request WITHOUT cookie. Envoy selects next available upstream host, generates cookie and attaches it to the response. From now on, the client attaches that cookie to each subsequent request to get stickiness. Currently, when upstream host encoded in the cookie is not available, Envoy selects next available upstream host and encodes it in the cookie return in the response. So the final result is the same.
The difference between cookie-based and header-based stateful sessions is that in header-based, encoding format of the header is documented and the client sort of takes over load balancing logic. In cookie-based stateful session, the format and encoding of the cookie are not documented and should not be used by user. 
IMHO `strict` settings only makes sense when user encodes the desired destination and says "I am only interested in that one destination" and such situation takes place only in header-based stateful session.

Reply:
> I believe that `strict` setting does not make sense for cookie-based SSA. Let's imagine that there is `strict` setting for cookie-based SSA. Envoy receives a request, determines that the desired host is not not available and returns 503. A client must send another request WITHOUT cookie. Envoy selects next available upstream host, generates cookie and attaches it to the response. From now on, the client attaches that cookie to each subsequent request to get stickiness. Currently, when upstream host encoded in the cookie is not available, Envoy selects next available upstream host and encodes it in the cookie return in the response. So the final result is the same.

I may be misunderstanding something here, but why wouldn't Envoy override the cookie without sending a 503 (just like it will override the header in this case)?
If the behavior is different between the two, I think this should be fixed.

> The difference between cookie-based and header-based stateful sessions is that in header-based, encoding format of the header is documented and the client sort of takes over load balancing logic. In cookie-based stateful session, the format and encoding of the cookie are not documented and should not be used by user. IMHO `strict` settings only makes sense when user encodes the desired destination and says "I am only interested in that one destination" and such situation takes place only in header-based stateful session.

IMHO in both cases the client is an "active" player that needs to provide some indication which session to attach the request to, but it is up to Envoy to set these values.

Reply:
I think that there is some disconnect here. While both mechanisms looks similar and even use similar code, my understanding is that use cases are different.

So, for cookie-based stateful session:
- client starts by sending a request without cookie
- Envoy selects upstream and responds to the client with a cookie (which consists of encoded upstream IP address)
- The client attaches the cookie to subsequent requests and Envoy guarantees that as long as the upstream is healthy, the requests will be routed there. Notice that the client is not supposed to decode the cookie. It does not know or care where the requests are routed. All it matters for the client is that requests always go to the same upstream host.
- When originally selected upstream is not available, Envoy selects a new upstream and cookie changes, but as long as the client attaches the new cookie the session will be sticky to the new upstream. 
Let's assume that there is `strict` option added to cookie-based stateful session. When the originally selected host becomes unavailable, the client receives 503. If the client keeps sending the same cookie, it will keep receiving 503. To get out of that situation, the client has to go back to the beginning - send a request without the cookie, Envoy will generate the new cookie and the client can start attaching it. We may argue that `strict` may make sense here, but since the client did not participate in the selection of upstream host, the meaning of `strict` does not make much sense IMHO.

Now, steps for for header-based stateful session are:
- the client MUST attach the header indicating where the request should be routed. The encoding of the IPAddress is well documented and forms an API between the client and Envoy. Notice that the client creates the token used for stateful session. In cookie-based, the token was created by Envoy.
- Envoy routes the request where the client asked to route it to.
- If the requested upstream host is not available, Envoy may select a new upstream and indicate that change to the client by attaching a header with the same name as the client used in the first request.
- But the client may require, that Envoy does not select a new upstream and send the request there, but only wants to know that upstream is not available because the client wants to select the new host itself. That is the meaning of `strict` setting for header-based stateful session.
 
@markdroth I may be missing something here, but I do not totally agree with your statement:

> The format looks essentially the same for header and cookie, and it's obviously designed to be opaque, not intended to be an API for the application to construct its own tokens for.

because the format is opaque to the client only in cookie-based scenario and is not opaque in header-based scenario. For header based, the client is active in encoding the stateful header, while in cookie-based scenario the client is only supposed to attach it to subsequent requests, without looking "inside" of the cookie.

OK, I am not totally adamant that `strict` should be only for header-based stickiness. Maybe it makes sense to have it for cookie-based scenario.

Reply:
I think you have the wrong impression of the intent of the header-based SSA mechanism.  The plugin itself [documents](https://github.com/envoyproxy/envoy/blob/0e441fbe53b4fe530ba41b8f24541afd5a734d86/api/envoy/extensions/http/stateful_session/header/v3/header.proto#L18) that the host selected by the LB policy will be returned to the client via a response header:

> This extension encodes the address of the upstream host selected by the load balancer into a response header with the header configuration.

I think the intended workflow here is exactly the same as with cookie-based SSA, the only difference being that the session state is passed via headers instead of cookies.  The expected workflow is that the data plane (e.g., Envoy or gRPC) will choose the host to use for the first request in the session, return the choice via a header, and the application needs to pass that header with all subsequent requests in the same session.  See #22475 for initial discussion of this feature.

With regard to the header value being documented, I think it was documented to show an example of how the mechanism works, not to imply that the format of the cookie or header is a public API that applications should use to independently generate the value.  Note that the cookie-based mechanism shows the [same example](https://github.com/envoyproxy/envoy/blob/0e441fbe53b4fe530ba41b8f24541afd5a734d86/api/envoy/extensions/http/stateful_session/cookie/v3/cookie.proto#L33).  I don't think you actually want to write an application that is tied to this format, because it is not guaranteed not to change over time.  In particular, I believe that there will soon be work done in Envoy to make SSA work with WeightedClusters, as was already done in gRPC (see [gRFC A60](https://github.com/grpc/proposal/blob/master/A60-xds-stateful-session-affinity-weighted-clusters.md)), and that will involve changing the header format.

Reply:
In fact, I am convinced by @cpakulski because of this sentence `When stateful session is based on cookies, there is assumption that cookie is generated by Envoy, not user.`.

Rather than format, I think another different point between header and cookie is that the cookie sometimes will be handled by the browser/client automatically but the session header should be handled by the user explicitly. In most case, I think the session cookie should be a **background role** and a strict mode that may result in unexpected 503 UH may bring noise to the users.

But I have no insistence on this point. It's OK for me to add this field the stateful session filter itself and document it clearly.

Reply:
So finally after reading @markdroth reply I think I know where the disconnect is. I believe that no-one is wrong. We just look from different perspectives. The original intent for header-based stateful session was discussed in #22475 and its original intent was to provide the the same functionality as cookie-based SSA, but using different header. The format of encoded information in cookie and header was originally the same. But it not the same any more - #26761 changed encoded format for cookie-based stateful session from text to protobuf. During work on that issue we received feedback from users that they use header to skip the initial random selection of upstream host. See https://github.com/envoyproxy/envoy/issues/26347#issuecomment-1503506078. That is why in the above discussion @wbpcode and myself sort of think about the header as something which is created by users, not Envoy. As described in https://github.com/envoyproxy/envoy/issues/30475#issue-1960168620, I see that users build use cases around the header-based stateful session. In that particular case, there is a feedback mechanism from backend (which upstream server has resources to accept a new client) and that IP address is encoded in the header and users do not want Envoy to select random upstream if the indicated is not available. The strict flag is geared towards only that use case, not the one where original client's request is sent without the header used for stateful session and Envoy generates one.

Reply:
My 2 cents: letting the downstream set a specific value to a header that is not part of some official standard (e.g., RFC), and assuming that a proxy will behave in a certain way is fragile, and could easily break things. It's somewhat similar to assuming that a call to a function with some value will always result in a side-effect that is not guaranteed per the function's spec.
Note that the contents of that response-header may change in the future, and the only guarantee is that the end-to-end process needs to work (receiving the response-header first and assigning it to future requests).

Reply:
Yes, I agree with you 100% @adisuissa.
To move things forward, I think we all agree that there is a need for such use case where a client indicates where the request should be routed on upstream side and stick to it in subsequent requests. We probably have only two choices:
- document the internal format of header-based stateful session and provide backwards compatibility if it changes in the future. Basically make it official API between client and Envoy's header-bases stateful session.
- implement it in a different way (additional filter?) and change documentation for header-based stateful session with note that the internal format should not be interpreted by clients and can change without any notice.

Any suggestions would be greatly appreciated.

Reply:
I wasn't aware of #26761, and I'm concerned about that change.  gRPC wasn't looped into this discussion at all, and we had intentionally implemented the same cookie format as Envoy, just to have consistent implementations.  Now the two implementations are different.  We'll need to consider whether or not this is important.

It's also really not clear to me that #26347 should have been addressed in Envoy to begin with.  I do think it's the application's responsibility to handle cookie lifetimes, and I think trying to duplicate this in the data plane is needless work.  It adds significant overhead to each individual request to have to decode an additional protobuf this way, and I don't think we should impose that overhead without sufficient offsetting benefit.  (The overhead here may be small compared to the overhead of running a sidecar proxy, but it's large compared to what users get with proxyless gRPC.)  The benefit seems small, considering that the client application should be dealing with the TTLs already.  It seems to me that this is basically adding a bunch of overhead just to work around a buggy client application, when we should instead have just fixed the client application.

Those issues aside, though, I think the broader question here is whether or not we intend the session state format to be a public API such that we support applications generating this data on their own.  I think that if we are going to treat this as a public API, we need to provide the same stability guarantees as the rest of xDS API.  This means that we would not be able to change the session state format in a non-backward-compatible way.  In other words, we would not be able to make a change like #26761 without continuing to support the old format as well, or else we'd risk breaking existing applications.  It will also cause problems when we make the changes to allow SSA to work with WeightedClusters.  How do we do that as a non-breaking change if the format is a public API?

In principle, I am also not convinced that there should be any fundamental difference between cookie-based and header-based session state in terms of whether the session state format is public.  I think the API being defined here is that the client application is responsible for maintaining the session state sent to it by the data plane (Envoy or gRPC).  The question of how it does that is up to the application; in principle, it's just as easy for the application to manage headers as it is to manage cookies.  I think the fact that one particular type of client (browsers) happens to have a built-in mechanism to manage cookies is essentially a coincidence; that is certainly not the case with gRPC, and we added cookie-based SSA support in gRPC for applications that will be managing the cookie directly.  If we say that the session state format is a public API, then there will be clients for whom it is just as easy to construct the initial cookie as it is to construct the initial header.  And the issues of needing to make changes to the session state format as we add new features apply equally to both mechanisms.

So I think the fundamental conflict here is between the desire to be able to change the session state format to add new features and the desire to make it a public API.  The only way I see to do both is to make the format extensible, which increases overhead (e.g., making it a protobuf to which we can add new fields).

I'd also like input on this from @ejona86 and @dfawley on the gRPC side.

Reply:
I'd be against this format being well-defined. It ossifies the feature. If it wouldn't cause other issues, I'd eagerly XOR it with a random number to avoid people touching it.

I'm not unsympathetic to someone wanting to steer a connection. Although that's really a different feature (is the client even saving the response cookie in that case?). Similarly if someone wanted to know which backend Envoy chose, that's not really what this is. That doesn't necessarily need a new filter, but it'd probably be fair to make it clearly different from the normal format (e.g., "api=1" magic string, different header name) so that the filter knows it is a different sort of input. It could read the header, do what's appropriate, and then still respond with a cookie with a fresh expiry or any other things. (As with Mark, it seems dubious to me to include the expiry in the cookie, but it seemed a fair example here of how you might take things from the well-defined API and enhance them in the session state.)

The last thing I'd want is "I can't upgrade to IPv6 because clients are assuming my backends are IPv4" and similar headaches, because clients got too creative.

Reply:
> So I think the fundamental conflict here is between the desire to be able to change the session state format to add new features and the desire to make it a public API. The only way I see to do both is to make the format extensible, which increases overhead (e.g., making it a protobuf to which we can add new fields).

I actually agree that we could create a new proto for stateful session that could be used for both cookie session and header session. And some simple flag could be used to distinguish the new proto and legacy raw IP. Like:

```
session-header: PROTO xxxxxxxxx
legacy-session-header: xxxxxxxxx
```

But I have no much awareness to the performance of proto parsing. At least it should be faster than json?

Reply:
Ha, I see the new format of gRPC in the https://github.com/grpc/proposal/blob/master/A60-xds-stateful-session-affinity-weighted-clusters.md . Seems Envoy also doesn't aware that. This is also a problem this need to be resolved.

```
206.12.3.6:8080;cluster:cloud-internal-istio:cloud_mp_635862331669_807571734533927564
```

Reply:
Created a issue to discuss the format of session header/cookie specifically. https://github.com/envoyproxy/envoy/issues/30678

Reply:
Thank you for all for your thoughts and suggestions. It seems that header-based SSA started as "cookie" under different name, but based on how the code and API was structured I was under impression that it was designed for different use cases (managed by client). Let's try to rectify this situation and design common, extendable format of the cookie/header and maybe separate mechanism to address use cases where a client wants to specify the upstream host.

Reply:
Thanks for creating that tracking issue, @wbpcode!  We can follow up on this discussion over there.

How shall we proceed with this PR?  Is strict mode still useful while the question of the session state being a public API is still undecided?  If not, then I guess we should hold off on this PR until we resolve that issue.  But if it is still useful, then how about we move this new field to the [`StatefulSession` proto](https://github.com/envoyproxy/envoy/blob/edf05524629732bf8f52c2ea18a94b207d5d3d72/api/envoy/extensions/filters/http/stateful_session/v3/stateful_session.proto#L20), so that it applies equally to both session state mechanisms?

Reply:
IMO, this strict mode and the format could be completely independent parts and could be implemented in different PR. So, I prefer to proceed with this PR.

> But if it is still useful, then how about we move this new field to the [StatefulSession proto](https://github.com/envoyproxy/envoy/blob/edf05524629732bf8f52c2ea18a94b207d5d3d72/api/envoy/extensions/filters/http/stateful_session/v3/stateful_session.proto#L20), so that it applies equally to both session state mechanisms?

Yeah, I agree.

Reply:
It makes sense to make it a part of [StatefulSession proto](https://github.com/envoyproxy/envoy/blob/edf05524629732bf8f52c2ea18a94b207d5d3d72/api/envoy/extensions/filters/http/stateful_session/v3/stateful_session.proto#L20). 

I agree that `strict` mode is separate and independent from the format. Just need to think about the use case when `strict` is enabled and the received cookie is expired. I guess that in that case, the stateful filter should immediately return 503 without selecting a new host.

Reply:
As I mentioned above, I would argue that we should not have tried to solve #26347 on the data plane side in the first place.  I think cookie expiration should be strictly an issue for the application to deal with, and the data plane shouldn't even know about the expiration time.  I think if the cookie is sent, the data plane should honor it, and `strict` mode should apply, just as it would be expected to if the cookie were not expired.

Reply:
@markdroth I think that in this case managing cookie's expiration time falls into grey area. I think that issuer of the cookie should verify expiration time. Normally it is a server.
So, here - on one side, it should not be part of data plane processing, but on the other side Envoy issued the cookie so Envoy should verify it's content.